flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -420,13 +429,23 @@ public void setInitialState(@Nullable JobManagerTaskRestore taskRestore) {
 				allPreviousExecutionGraphAllocationIds,
 				allocationTimeout);
 
-			// IMPORTANT: We have to use the synchronous handle operation (direct executor) here so
-			// that we directly deploy the tasks if the slot allocation future is completed. This is
-			// necessary for immediate deployment.
-			final CompletableFuture<Void> deploymentFuture = allocationFuture.thenAccept(
-				(FutureConsumerWithException<Execution, Exception>) value -> deploy());
+			final ComponentMainThreadExecutor mainThreadExecutor =
+				vertex.getExecutionGraph().getJobMasterMainThreadExecutor();
+			FutureConsumerWithException<Execution, Exception> deployment = value -> deploy();
+			final CompletableFuture<Void> deploymentFuture;
+			if (allocationFuture.isDone()) {
+				deploymentFuture = allocationFuture.thenAccept(deployment);
+			} else if (queued) {
+				deploymentFuture = allocationFuture.thenAcceptAsync(
+					deployment, mainThreadExecutor);
```

Comment:
I think that would rarely make a difference, because we alrady checked `allocationFuture.isDone()` right before in L 436 at the beginning of the `if`block.

Reply:
With our change in Execution, L 555 that is correct.